File: com.trollworks.gcs/src/com/trollworks/gcs/utility/units/LengthUnits.java
/*
 * Copyright ©1998-2020 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.utility.units;

import com.trollworks.gcs.utility.Fixed6;
import com.trollworks.gcs.utility.I18n;

import java.text.MessageFormat;

/** Common length units. */
public enum LengthUnits implements Units {
    /** Points (1/72 of an inch). */
    PT(Fixed6.ONE.div(new Fixed6(72)), false) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Points");
        }
    },
    /** Inches. */
    IN(Fixed6.ONE, false) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Inches");
        }
    },
    /** Feet. */
    FT(new Fixed6(12), false) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Feet");
        }
    },
    /** Feet and Inches */
    FT_IN(Fixed6.ONE, false) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Feet & Inches");
        }

        @Override
        public String toString() {
            return I18n.Text("Feet (') & Inches (\")");
        }

        @Override
        public String format(Fixed6 value, boolean localize) {
            Fixed6 twelve = new Fixed6(12);
            Fixed6 feet   = value.div(twelve).trunc();
            Fixed6 inches = value.sub(twelve.mul(feet));
            if (feet.greaterThan(Fixed6.ZERO)) {
                String buffer = formatNumber(feet, localize) + "'";
                if (inches.greaterThan(Fixed6.ZERO)) {
                    return buffer + ' ' + formatNumber(inches, localize) + '"';
                }
                return buffer;
            }
            return formatNumber(value, localize) + '"';
        }

        private String formatNumber(Fixed6 value, boolean localize) {
            return localize ? value.toLocalizedString() : value.toString();
        }
    },
    /** Yards. */
    YD(new Fixed6(36), false) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Yards");
        }
    },
    /** Miles. */
    MI(new Fixed6(5280).mul(new Fixed6(12)), false) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Miles");
        }
    },
    /** Millimeters. */
    MM(new Fixed6("0.1", Fixed6.ZERO, false).div(LengthValue.METRIC_CONVERSION_FACTOR), true) { // entered as text to ensure precision
        @Override
        public String getLocalizedName() {
            return I18n.Text("Millimeters");
        }
    },
    /** Centimeters. */
    CM(Fixed6.ONE.div(LengthValue.METRIC_CONVERSION_FACTOR), true) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Centimeters");
        }
    },
    /** Kilometers. */
    KM(new Fixed6(100000).div(LengthValue.METRIC_CONVERSION_FACTOR), true) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Kilometers");
        }
    },
    /** Meters. Must be after all the other 'meter' types. */
    M(new Fixed6(100).div(LengthValue.METRIC_CONVERSION_FACTOR), true) {
        @Override
        public String getLocalizedName() {
            return I18n.Text("Meters");
        }
    };

    private Fixed6  mFactor;
    private boolean mIsMetric;

    LengthUnits(Fixed6 factor, boolean isMetric) {
        mFactor = factor;
        mIsMetric = isMetric;
    }

    @Override
    public Fixed6 convert(Units units, Fixed6 value) {
        return units.getFactor().mul(value).div(mFactor);
    }

    @Override
    public Fixed6 normalize(Fixed6 value) {
        return mFactor.mul(value);
    }

    @Override
    public Fixed6 getFactor() {
        return mFactor;
    }

    @Override
    public String format(Fixed6 value, boolean localize) {
        return MessageFormat.format("{0} {1}", localize ? value.toLocalizedString() : value.toString(), getAbbreviation());
    }

    @Override
    public LengthUnits[] getCompatibleUnits() {
        return values();
    }

    @Override
    public String getAbbreviation() {
        return name().toLowerCase();
    }

    @Override
    public String toString() {
        return String.format("%s (%s)", getLocalizedName(), getAbbreviation());
    }

    public boolean isMetric() {
        return mIsMetric;
    }
}

File: model/fxp/length_units.go
/*
 * Copyright ©1998-2023 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package fxp

import (
	"strings"
)

// Format the length for this LengthUnits.
func (enum LengthUnits) Format(length Length) string {
	inches := Int(length)
	switch enum {
	case FeetAndInches:
		oneFoot := From(12)
		feet := inches.Div(oneFoot).Trunc()
		inches -= feet.Mul(oneFoot)
		if feet == 0 && inches == 0 {
			return "0'"
		}
		var buffer strings.Builder
		if feet > 0 {
			buffer.WriteString(feet.String())
			buffer.WriteByte('\'')
		}
		if inches > 0 {
			buffer.WriteString(inches.String())
			buffer.WriteByte('"')
		}
		return buffer.String()
	case Inch:
		return inches.String() + " " + enum.Key()
	case Feet:
		return inches.Div(From(12)).String() + " " + enum.Key()
	case Yard, Meter:
		return inches.Div(From(36)).String() + " " + enum.Key()
	case Mile:
		return inches.Div(From(63360)).String() + " " + enum.Key()
	case Centimeter:
		return inches.Div(From(36)).Mul(From(100)).String() + " " + enum.Key()
	case Kilometer:
		return inches.Div(From(36000)).String() + " " + enum.Key()
	default:
		return FeetAndInches.Format(length)
	}
}

// ToInches converts the length in this LengthUnits to inches.
func (enum LengthUnits) ToInches(length Int) Int {
	switch enum {
	case FeetAndInches, Inch:
		return length
	case Feet:
		return length.Mul(From(12))
	case Yard:
		return length.Mul(From(36))
	case Mile:
		return length.Mul(From(63360))
	case Centimeter:
		return length.Mul(From(36)).Div(From(100))
	case Kilometer:
		return length.Mul(From(36000))
	case Meter:
		return length.Mul(From(36))
	default:
		return FeetAndInches.ToInches(length)
	}
}
